### ISSUE TITLE
Performance with more than 20 windows open

### ISSUE THREAD

**Issue by dbjergaard** [stumpwm/stumpwm#382]
If there are more than 20 windows open in a group, the StumpWM UI lags significantly in responding to events and redrawing frames.  

Naively I expect that these two things should be independent of the number of windows in a group.  If there are 20 windows, only one or two are visible, and they should be the only ones that are re-drawn or resized when the frames move or change.  

Steps to reproduce:
- open "a bunch" of terminal emulators (where a bunch if more than 20 and less than 100, depending on the beefiness of your system)
- Interact with StumpWM, split the frame, hit the prefix key, change focus.  There should be a noticeable lag.  

To fix:
research how to profile lisp code, setup the situation above, measure and identify hot spots.  Try to only update what is visible to the screen, and allow other windows to sit dormant until they're explictly raised to the front of a frame.

**Comment by dbjergaard** [comment]
The mode-line must be drawn for the lag to show up, disabling the mode-line increases the performance significantly.

**Comment by TeMPOraL** [comment]
Can't reproduce. Is this still happening? I'm willing to spend some time profiling StumpWM to figure this out.

**Comment by dbjergaard** [comment]
yes, although it is slightly weird, I don't see the drawing lag that I saw before, but I'm definitely getting input lag. This could also be related to #232.

**Comment by TeMPOraL** [comment]
Did you, by any chance, simplify your modeline?

Last night I was trying to figure out why `IRESIZE` lags for me, and it turned out it redraws modeline at each step, and my configuration caused it to trigger an external shell command on each such redraw. With modeline config changed to not do any long operations, it doesn't have any noticeable lag. Similarly #252 doesn't affect scrolling much with computationally-lean modeline.

**Comment by equwal** [comment]
> research how to profile lisp code, setup the situation above, measure and identify hot spots. 
SBCL has two profilers built in for you to use: a statistical one and a deterministic one. They are quite easy to use. [SBCL Manual Section 16, Profiling](http://www.sbcl.org/manual/#Profiling).
I don't have any lag opening dozens of `xterm` instances.

**Comment by dmb2** [comment]
This is still a problem when > 50 windows is open.

Specifically, the modeline must be enabled when lots of windows are created.  If you disable the modeline, spawn 50-100 windows, and enable the modeline, lag doesn't seem to occur.

If you spawn 50-100 windows with the modeline enabled stumpwm will slow to a crawl. 

If the mode-line string was limited to the screen-width in pixels it the issue might be circumvented.  Right now mode-line.lisp does a mapcar across all elements in the modeline and accumulates the results as a string.  If the string is limited to what can be drawn on the screen, performance may improve.
